components/omega/doc/design/Metadata.md

```diff
@@ -0,0 +1,320 @@
+be assigned the appropriate R4 or R8 by the compiler through
+the correct choice of aliased interfaces in the methods below.
+
+    class Metadata{
+
+       private:
+           std::string name; // field name to which metadata attached
+                             // use reserved names "code", "simulation"
+                             // or the I/O stream name for general
+                             // metadata associated with those
+
+           I4 numI4;   // number of I4 metadata entries
+           I4 numI8;   // number of I8 metadata entries
+           I4 numR4;   // number of R4 metadata entries
+           I4 numR8;   // number of R8 metadata entries
+           I4 numBool; // number of Boolean metadata entries
+           I4 numStr;  // number of string metadata entries
+
+           // Metadata name,value pairs for each type
+           std::map<std::string, I4> metaMapI4;
+           std::map<std::string, I8> metaMapI8;
+           std::map<std::string, R4> metaMapR4;
+           std::map<std::string, R8> metaMapR8;
+           std::map<std::string, bool> metaMapBool;
+           std::map<std::string, std::string> metaMapStr;
+
+           // Dimensions - only for defining fields/variables
+           I4 nDims;   // number of dimensions
+           std::vector<MetadataDim*> dims; // pointer to the defined
+                                           // dimension for each dim
+
+       public:
+          [methods described below]
+    };
+
+A second class is needed to describe dimensions for arrays
+and vectors. It is basically only a name, length pair to define
+the length of each dimension.
+
+    class MetadataDim{
+
+       private:
+          std::string name; // name of dimension
+          I4 length;        // length of dimension
+
+       public:
+          [methods described below]
+    };
+
+### 4.2 Methods
+
+#### 4.2.1 Metadata Constructors and Destructors
+
+There will be two constructors for the metadata class. In addition
+to initializing the metadata, these constructors will also add
+the metadata to the list of defined fields so will also check that
+the metadata has not already be defined and fail if already exists.
+
+    // construct empty instance by name, primarily for global
+    // metadata not assigned to a variable/field
+    Metadata(const std::string name); // name to assign
+
+    // construct metadata for a variable/field using
+    // required metadata. Not that if stdName doesn't exist
+    // an empty string can be supplied
+    Metadata(const std::string name,        // name of var
+             const std::string description, // long name or description
+             const std::string units,       // units
+             const std::string stdName,     // CF standard name
+             const int         nDims,       // number of dimensions
+             const std::vector<MetadataDims*> // vector of dim pointers
+             );
+
+Do we want to require any other metadata (valid min/max, fill value)?
```

Comment:
See above comments about fill value masking. 

I think that including valid min/max metadata could be useful, that is, they could be used directly in the new version of `ocn_validate_state` to check whether values are in bounds.

Reply:
add valid min/max as required metadata. see above comment about fill value and masking